peer.go

```diff
@@ -1,516 +1,518 @@
-	// is received.
-	p.relayMtx.Lock()
-	p.disableRelayTx = msg.DisableRelayTx
-	p.relayMtx.Unlock()
-
-	// Inbound connections.
-	if p.inbound {
-		// Set up a NetAddress for the peer to be used with AddrManager.
-		// We only do this inbound because outbound set this up
-		// at connection time and no point recomputing.
-		na, err := newNetAddress(p.conn.RemoteAddr(), p.services)
-		if err != nil {
-			p.logError("Can't get remote address: %v", err)
-			p.Disconnect()
+		// Ignore old style addresses which don't include a timestamp.
+		if p.ProtocolVersion() < wire.NetAddressTimeVersion {
 			return
 		}
-		p.na = na
 
-		// Send version.
-		err = p.pushVersionMsg()
-		if err != nil {
-			p.logError("Can't send version message to %s: %v",
-				p, err)
+		// A message that has no addresses is invalid.
+		if len(msg.AddrList) == 0 {
+			p.LogError("Command [%s] from %s does not contain any addresses",
+				msg.Command(), p)
 			p.Disconnect()
 			return
 		}
-	}
-
-	// Send verack.
-	p.QueueMessage(wire.NewMsgVerAck(), nil)
 
-	// Update the address manager and request known addresses from the
-	// remote peer for outbound connections.  This is skipped when running
-	// on the simulation test network since it is only intended to connect
-	// to specified peers and actively avoids advertising and connecting to
-	// discovered peers.
-	if !cfg.SimNet {
-		p.updateAddresses(msg)
-	}
+		for _, na := range msg.AddrList {
+			// Don't add more address if we're disconnecting.
+			if !p.Connected() {
+				return
+			}
 
-	// Add the remote peer time as a sample for creating an offset against
-	// the local clock to keep the network time in sync.
-	p.server.timeSource.AddTimeSample(p.addr, msg.Timestamp)
+			// Set the timestamp to 5 days ago if it's more than 24 hours
+			// in the future so this address is one of the first to be
+			// removed when space is needed.
+			now := time.Now()
+			if na.Timestamp.After(now.Add(time.Minute * 10)) {
+				na.Timestamp = now.Add(-1 * time.Hour * 24 * 5)
+			}
 
-	// Signal the block manager this peer is a new sync candidate.
-	p.server.blockManager.NewPeer(p)
+			// Add address to known addresses for this peer.
+			p.KnownAddresses[addrmgr.NetAddressKey(na)] = struct{}{}
+		}
 
-	// TODO: Relay alerts.
+		// Add addresses to server address manager.  The address manager handles
+		// the details of things such as preventing duplicate addresses, max
+		// addresses, and last seen updates.
+		// XXX bitcoind gives a 2 hour time penalty here, do we want to do the
+		// same?
+		s.addrManager.AddAddresses(msg.AddrList, p.NA())
+	}
 }
 
 // pushTxMsg sends a tx message for the provided transaction hash to the
 // connected peer.  An error is returned if the transaction hash is not known.
-func (p *peer) pushTxMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
+var pushTxMsg = func(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```

Comment:
Any reason this is a variable?

```Go
func pushTxMsg(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
would allow the compiler to be more efficient.